students/lauraannf/lessons/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,171 @@
+                                          'phone_number': row['Phone Number'],
+                                          'status': row['Status'],
+                                          'credit_limit': row['Credit Limit']})
+                try:
+                    customers.insert_many(customer_list)
+                except Exception as ex:
+                    LOGGER.warning(ex)
+        except Exception as ex:
+            LOGGER.warning(ex)
+            LOGGER.warning('error when opening customer file')
+        customer_count_new = database.customers.count_documents({})
+
+        queue.put((customer_count_new - customer_count_orig, customer_count_orig,
+               customer_count_new, time.clock() - start_c))
+
+def import_rentals(directory_name, rental_file, queue):
+    """imports data from csv files and puts in database"""
+    start_r = time.clock()
+    mongo = MongoDBConnection()
+
+    with mongo:
+        database = mongo.connection.HPNorton
+        rental_count_orig = database.rentals.count_documents({})
+        rentals = database["rentals"]
+        rental_list = []
+
+        try:
+            with open(os.path.join(directory_name, rental_file)) as rental_csv:
+                csv_reader = csv.DictReader(rental_csv)
+
+
+                for row in csv_reader:
+                    rental_list.append({'rental_id': row['Rental ID'],
+                                        'customer_id': row['Customer ID'],
+                                        'product_id': row['Product ID'],
+                                        'quantity': row['Quantity']})
+                try:
+                    rentals.insert_many(rental_list)
+                except Exception as ex:
+                    LOGGER.warning(ex)
+        except Exception as ex:
+            LOGGER.warning(ex)
+            LOGGER.warning('error when opening rental file')
+        rental_count_new = database.rentals.count_documents({})
+
+        queue.put((rental_count_new - rental_count_orig, rental_count_orig,
+               rental_count_new, time.clock() - start_r))
+
+
+def delete_database():
+    """deletes database to start fresh"""
+    mongo = MongoDBConnection()
+
+    with mongo:
+        database = mongo.connection.HPNorton
+
+        customers = database["customers"]
+        rentals = database["rentals"]
+        products = database["products"]
+
+        customers.drop()
+        rentals.drop()
+        products.drop()
+
+
+
+if __name__ == "__main__":
+    start = time.clock()
+    queue = Queue()
+    queue_results = []
+    threads = [Thread(target=import_products, args=('csvfiles','inventory.csv', queue)),
+               Thread(target=import_customers, args=('csvfiles', 'customers.csv', queue)), 
+               Thread(target=import_rentals, args=('csvfiles', 'rental.csv', queue ))]
+    for thread in threads:
+        thread.start()
+        queue_results.append(queue.get())
```

Comment:
Consider that _queue_results.append(queue.get())_ will not wait for the thread you just started with _thread.start()_ to finish.

You should follow a sequence similar to this one:

1. Launch all threads.
2. Go through the list of threads, executing a _thread.join()_ which will effectively wait for that thread to finish (other threads can continue to run without issues while this happens).
3. Pull the data from the queue.

The code will look like this:

```
    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join() 
        queue_results.append(queue.get())
```